lib/resty/aes.lua

```diff
@@ -104,6 +107,18 @@ cipher = function (size, _cipher)
 end
 _M.cipher = cipher
 
+local function get_error()
+    local errno = C.ERR_get_error()
+    if errno == 0 then
+        return nil
+    end
+
+    local msg = ffi_new("char[?]", 256)
+    C.ERR_error_string_n(errno, msg, 256)
```

Comment:
One OpenSSL call may cause several error messages. Therefore, to handle the error message properly, we need to draw the whole error messages out like:
https://github.com/spacewander/lua-resty-rsa/blob/6a6073f2dcfc319fdf760c65a26382f2a8cf67df/lib/resty/rsa.lua#L158

Reply:
> One OpenSSL call may cause several error messages

I suspected it but couldn't find confirmation in the OpenSSL documentation. Does the documentation say it clearly or you know it from source code (or another source)?

Reply:
attention with `https://github.com/spacewander/lua-resty-rsa/blob/6a6073f2dcfc319fdf760c65a26382f2a8cf67df/lib/resty/rsa.lua#L158`.
`C.ERR_reason_error_string` function may return a null value when the code not match a human readable message.
so null value checking for err variable
```
local err = C.ERR_reason_error_string(code)
if err ~= ffi.null then
    err_queue[i] = ffi_str(err)
    ...
end
```

Reply:
@toruneko 
Yes, you are right. We need to check the returned err before using it.

Reply:
@un-def 
Here is the doc:
https://en.wikibooks.org/wiki/OpenSSL/Error_handling

Reply:
BTW, sometimes OpenSSL don't even return an error message when the call is failed. (Perhaps it is a bug): https://github.com/openssl/openssl/issues/11962

So strictly speaking, we need to provide a default one if no error is found:
https://github.com/spacewander/lua-resty-rsa/blob/6a6073f2dcfc319fdf760c65a26382f2a8cf67df/lib/resty/rsa.lua#L180

Reply:
@spacewander I've seen it, but this is not the official OpenSSL wiki and has not been updated for a long time. I meant the official docs hosted at https://www.openssl.org/ or official wiki https://wiki.openssl.org/ Anyway, I agree that it's worth popping all entries from the error stack, it's safe and future-compatible.